lens/fetcher.go

```diff
@@ -0,0 +1,400 @@
+		if !ok {
+			continue
+		}
+		docAsMap[fieldName] = fieldValue
+	}
+
+	docAsMap[request.DeletedFieldName] = doc.Status.IsDeleted()
+
+	return docAsMap, nil
+}
+
+// lensDocToCoreDoc converts a LensDoc to a core.Doc.
+func (f *lensedFetcher) lensDocToCoreDoc(mapping *core.DocumentMapping, docAsMap LensDoc) (core.Doc, error) {
+	doc := mapping.NewDoc()
+
+	for fieldName, fieldByteValue := range docAsMap {
+		if fieldName == request.KeyFieldName {
+			key, ok := fieldByteValue.(string)
+			if !ok {
+				return core.Doc{}, core.ErrInvalidKey
+			}
+
+			doc.SetKey(key)
+			continue
+		}
+
+		fieldDesc, fieldFound := f.fieldDescriptionsByName[fieldName]
+		if !fieldFound {
+			// Note: This can technically happen if a Lens migration returns a field that
+			// we do not know about. In which case we have to skip it.
+			continue
+		}
+
+		fieldValue, err := core.DecodeFieldValue(fieldDesc, fieldByteValue)
+		if err != nil {
+			return core.Doc{}, err
+		}
+
+		index := mapping.FirstIndexOfName(fieldName)
+		doc.Fields[index] = fieldValue
+	}
+
+	if value, ok := docAsMap[request.DeletedFieldName]; ok {
+		if wasDeleted, ok := value.(bool); ok {
+			if wasDeleted {
+				doc.Status = client.Deleted
+			} else {
+				doc.Status = client.Active
+			}
+		}
+	}
+
+	doc.SchemaVersionID = f.col.Schema.VersionID
+
+	return doc, nil
+}
+
+// lensDocToClientDoc converts a LensDoc to a client.Document.
+func (f *lensedFetcher) lensDocToClientDoc(docAsMap LensDoc) (*client.Document, error) {
+	key, err := client.NewDocKeyFromString(docAsMap[request.KeyFieldName].(string))
+	if err != nil {
+		return nil, err
+	}
+	doc := client.NewDocWithKey(key)
+
+	for fieldName, fieldByteValue := range docAsMap {
+		if fieldName == request.KeyFieldName {
+			continue
+		}
+
+		fieldDesc, fieldFound := f.fieldDescriptionsByName[fieldName]
+		if !fieldFound {
+			// Note: This can technically happen if a Lens migration returns a field that
+			// we do not know about. In which case we have to skip it.
+			continue
+		}
+
+		fieldValue, err := core.DecodeFieldValue(fieldDesc, fieldByteValue)
+		if err != nil {
+			return nil, err
+		}
+
+		err = doc.SetAs(fieldDesc.Name, fieldValue, fieldDesc.Typ)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	doc.SchemaVersionID = f.col.Schema.VersionID
+
+	// Note: client.Document does not have a means of flagging as to whether it is
+	// deleted or not, and, currently the fetcher does not ever returned deleted items
+	// from the function that returs this type.
+
+	return doc, nil
+}
+
+// updateDataStore updates the datastore with the migrated values.
+//
+// This removes the need to migrate a document everytime it is fetched as the second time around
+// the underlying fetcher will return the migrated values cached in the datastore instead of the
+// underlying dag store values.
+func (f *lensedFetcher) updateDataStore(ctx context.Context, original map[string]any, migrated map[string]any) error {
```

Comment:
todo: use `LensDoc` alias

```go
func (f *lensedFetcher) updateDataStore(ctx context.Context, original LensDoc, migrated LensDoc) error {
```

Reply:
Why?  Within the context of this function the fact that they happen to be LensDocs is irrelevant and accidental.

Reply:
well, it's just an alias, not a big deal I guess. But it would be consistent if you chose this term to convey a meaning.
The `sourceLensDoc` is of type `LensDoc` as `coreDocToLensDoc`
```go
sourceLensDoc, err := coreDocToLensDoc(mapping, doc)
...
migratedLensDoc, err := f.lens.Value()
...
err = f.updateDataStore(ctx, sourceLensDoc, migratedLensDoc)
```
here `coreDocToLensDoc` returns `LensDoc` and `f.lens.Value()` returns `LensDoc` as `f.lens` implements interface `enumerable.Enumerable[LensDoc]`. So I don't see any accident here.

Reply:
Agree with Islam here, the alias is specifically added to convey the type meaning, might as well use it consistently, otherwise I don't see much point in having the alias.

Reply:
`updateDataStore` takes a map, it does not care what that map was used for previously.  The only reason it takes a map is out of convenience because of the 3 annoying functions in the Fetcher that prevent it from using the doc types (without  duplication).

I think you guys are looking for meaning where there is none.

Reply:
I think you need to see the alias as a concrete type (or a larger interface), and `map[string]any` as an interface.  It is common to pass around a concrete type into functions taking an interface.

The main code body cares that it is a LensDoc.  This function does not.

Reply:
> I think you need to see the alias as a concrete type (or a larger interface), and map[string]any as an interface.

This gets into extreamly murky waters, especially in the world of traditional/expected Go, because I have come across a lot of code bases with alias's and never thought of them using this relationship. 

At the moment, nothing cares if its a lens doc or a `map[string]any`, as its  just an alias. In Go, theres only two reasons to alias a type A) for maintainability if you plan to overhaul that type, you can alias so as to reduce future maintenence B) For semantics (wanting to convey meaning that a primitive type is used in a certain condition/manner). 

From what I can tell, your use of aliasing is (B). We're "updating the datastore" with the source doc, which you make an explicit effort to "convert" to a `lensDoc` using a function that returns `LensDoc`, not to a `map[string]any` (even tho theyre aliased), and a migrated lens doc, which comes from a `Enumerable[LensDoc]` type. So again, you go thru all the effort of ensuring all these types use the defined alias, but then (seemingly arbitrarily) use the internal type youre aliasing. 

Again, just oddly inconsistent, and enough for both Islam and me pause for a moment and work backwards thru the call stack to verify this is intended.

PS. Still soft preference, not a blocker, but I do hope you understand the reasonsing/intent in my comment 👍.

@islamaliev any thoughts/strong preference?